api/api.go

```diff
@@ -6,40 +6,155 @@ import (
+// authentication check. The wrapped handler is only called if the caller is authenticated
+func (api *API) isAuthenticated(handler http.HandlerFunc) http.HandlerFunc {
+	return dphandlers.CheckIdentity(handler)
+}
+
+// isAuthorised wraps a http.HandlerFunc another http.HandlerFunc that checks the caller is authorised to perform the
+// requested action. required is the permissions required to perform the action, handler is the http.HandlerFunc to
+// apply the check to. The wrapped handler is only called if the caller has the required permissions.
+func (api *API) isAuthorised(required auth.Permissions, handler http.HandlerFunc) http.HandlerFunc {
+	return api.permissions.Require(required, handler)
+}
+
+// isAuthorised wraps a http.HandlerFunc another http.HandlerFunc that checks the caller is authorised to perform the
+// requested topics action. This authorisation check is specific to topics. required is the permissions required to
+// perform the action, handler is the http.HandlerFunc to apply the check to. The wrapped handler is only called if the
+// caller has the required topic permissions.
+func (api *API) isAuthorisedForTopics(required auth.Permissions, handler http.HandlerFunc) http.HandlerFunc {
+	return api.topicPermissions.Require(required, handler)
+}
+
+// get register a GET http.HandlerFunc.
+func (api *API) get(path string, handler http.HandlerFunc) {
+	api.Router.HandleFunc(path, handler).Methods("GET")
+}
+
+// get register a PUT http.HandlerFunc.
+func (api *API) put(path string, handler http.HandlerFunc) {
+	api.Router.HandleFunc(path, handler).Methods("PUT")
+}
+
+// get register a POST http.HandlerFunc.
+func (api *API) post(path string, handler http.HandlerFunc) {
+	api.Router.HandleFunc(path, handler).Methods("POST")
+}
+
+// get register a DELETE http.HandlerFunc.
+func (api *API) delete(path string, handler http.HandlerFunc) {
+	api.Router.HandleFunc(path, handler).Methods("DELETE")
+}
+
+func (api *API) authenticate(r *http.Request, logData log.Data) bool {
+	var authenticated bool
+
+	if api.enablePrivateEndpoints {
+		var hasCallerIdentity, hasUserIdentity bool
+
+		// NOTE:
+		// If the identity exists then the user has been authenticated.
+		// There is an earlier step in the middleware which will call off to zebedee to
+		// authenticate the request (user/service) and this will add the identity to the
+		// request context for later use in the application ...
+		// ... which happens to be here:
+
+		callerIdentity := dprequest.Caller(r.Context())
+		if callerIdentity != "" {
+			logData["caller_identity"] = callerIdentity
+			hasCallerIdentity = true
+		}
+
+		userIdentity := dprequest.User(r.Context())
+		if userIdentity != "" {
+			logData["user_identity"] = userIdentity
+			hasUserIdentity = true
+		}
+
+		if hasCallerIdentity || hasUserIdentity {
+			authenticated = true
+		}
+		logData["authenticated"] = authenticated
+
+		return authenticated
```

Comment:
you could probably omit this return and let it fall through as it will return immediately anyway. No functional different but 1 less line of code.

Reply:
Done.
(i copied that function from dp-dataset-api and corrected it after much discussion / agreement with Nathan)
So, it looks like that improvement was missed in the original work.